taichi/ir/analysis.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "taichi/ir/ir.h"
+#include <atomic>
+#include <unordered_set>
+#include <unordered_map>
+
+TLANG_NAMESPACE_BEGIN
+
+class DiffRange {
+ private:
+  bool related;
+
+ public:
+  int coeff;
+  int low, high;
+
+  DiffRange() : DiffRange(false, 0) {
+  }
+
+  DiffRange(bool related, int coeff) : DiffRange(related, 0, 0) {
+    TI_ASSERT(related == false);
+  }
+
+  DiffRange(bool related, int coeff, int low)
+      : DiffRange(related, coeff, low, low + 1) {
+  }
+
+  DiffRange(bool related, int coeff, int low, int high)
+      : related(related), coeff(coeff), low(low), high(high) {
+    if (!related) {
+      this->low = this->high = 0;
+    }
+  }
+
+  bool related_() const {
+    return related;
+  }
+
+  bool linear_related() const {
+    return related && coeff == 1;
+  }
+
+  bool certain() {
+    TI_ASSERT(related);
+    return high == low + 1;
+  }
+};
+
+// IR Analysis
+namespace irpass {
```

Comment:
We can use `namespace irpass::analysis` here.

Reply:
Will this cause an error if `namespace irpass` not defined yet?